### PR TITLE
Fix tests to ignore the startup file

### LINKED ISSUE #17642: Tests don't ignore .juliarc.jl
On master, I have the following error show up with `make test`, and the tests hang.  If I move my .juliarc.jl out of the way, the tests pass as expected.

I believe there are a number of places in the tests that the `--startup-file=no` option is missing.  I have found many of them, which allows those tests to proceed, but some are still having issues. If I manage to figure it out, I'll set up a PR.

```
ERROR: LoadError: `@osx` is deprecated, use `@static is_apple()` instead
 in include_from_node1(::String) at ./loading.jl:426
 in try_include at ./client.jl:200 [inlined]
 in load_juliarc() at ./client.jl:278
 in process_options(::Base.JLOptions) at ./client.jl:216
 in _start() at ./client.jl:318
while loading /Users/rrock/.juliarc.jl, in expression starting on line 6
```

### PR DESCRIPTION
**Pull request by rsrock**
To find the offending tests, it was helpful to use
a .juliarc.jl with just `error("don't include me")`.

Fixes #17642

### CODE REVIEW

File: test/cmdlineargs.jl [hunk 1]
```diff
@@ -277,21 +270,33 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
     end
 end
 
+let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # --startup-file
+    # Note: cannot cleanly test for --startup-file=yes, because errors or deprication
+    # warnings in ~/.juliarc.jl cause the tests to freeze
+    let JL_OPTIONS_STARTUPFILE_ON = 1,
+        JL_OPTIONS_STARTUPFILE_OFF = 2
+        # @test parse(Int,readchomp(`$exename -E "Base.JLOptions().startupfile" --startup-file=yes`)) == JL_OPTIONS_STARTUPFILE_ON
```
Comment (rsrock) [inline test/cmdlineargs.jl]: This is a bit of a problem.  I don't know how to properly test that the startup file is being used, if a broken startup file causes the tests to hang.

File: test/cmdlineargs.jl [hunk 2]
```diff
@@ -69,6 +69,7 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
     end
 
     # --procs
+    # FIXME: is the -p 2 option incompatible with --startup-file=no ?
     @test readchomp(`$exename -q -p 2 -e "println(nworkers())"`) == "2"
```
Comment (rsrock) [inline test/cmdlineargs.jl]: Here's a problem:

```❯ ./julia -q --startup-file=no -e "println(nworkers())"
1

rrock at Uluru in ~/Documents/Code/julia5  rsr/test-ignore-startup
❯ ./julia -q -p 2 --startup-file=no -e "println(nworkers())"
ERROR: LoadError: Don't include me!
 in error(::String) at ./error.jl:21
 in include_from_node1(::String) at ./loading.jl:426
 in try_include at ./client.jl:200 [inlined]
 in load_juliarc() at ./client.jl:278
 in process_options(::Base.JLOptions) at ./client.jl:216
 in _start() at ./client.jl:318
while loading /Users/rrock/.juliarc.jl, in expression starting on line 1
ERROR: LoadError: Don't include me!
 in error(::String) at ./error.jl:21
 in include_from_node1(::String) at ./loading.jl:426
 in try_include at ./client.jl:200 [inlined]
 in load_juliarc() at ./client.jl:278
 in process_options(::Base.JLOptions) at ./client.jl:216
 in _start() at ./client.jl:318
while loading /Users/rrock/.juliarc.jl, in expression starting on line 1

[hangs here]
```

It looks like the `-p 2` option ignores the `--startup-file=no` option.  Should I file a separate issue for this?

File: test/cmdlineargs.jl (hunk 1 above)
Comment (vtjnash) [inline test/cmdlineargs.jl]: Set the HOME env variable to a controlled location?

File: test/cmdlineargs.jl (hunk 2 above)
Comment (vtjnash) [inline test/cmdlineargs.jl]: Yes

File: test/cmdlineargs.jl [hunk 3]
```diff
@@ -277,21 +270,33 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
     end
 end
 
+let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # --startup-file
+    # Note: cannot cleanly test for --startup-file=yes, because errors or deprication
```
Comment (tkelman) [inline test/cmdlineargs.jl]: deprecation

**Comment by rsrock** [comment]
Ok, I addressed the comments and filed a separate issue for `julia -p`.  I have one remaining question about setting the HOME variable, because I doubt what I did here will work on Windows.

File: test/cmdlineargs.jl [hunk 4]
```diff
@@ -277,21 +269,32 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
     end
 end
 
+let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # --startup-file
+    let JL_OPTIONS_STARTUPFILE_ON = 1,
+        JL_OPTIONS_STARTUPFILE_OFF = 2
+        # `HOME=/tmp` is to bypass any errors in user startup files, which cause the tests to hang
+        @test parse(Int,readchomp(`HOME=/tmp $exename -E "Base.JLOptions().startupfile" --startup-file=yes`)) == JL_OPTIONS_STARTUPFILE_ON
```
Comment (tkelman) [inline test/cmdlineargs.jl]: you'll probably need to use the `setenv` function for this

Comment (rsrock) [inline test/cmdlineargs.jl]: Aha, the julia `setenv` function.  That helps.

**Comment by rsrock** [comment]
The Travis failure looks unrelated.  I think this part is ready to go.

File: test/cmdlineargs.jl [hunk 5]
```diff
@@ -277,21 +269,32 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes`
     end
 end
 
+let exename = `$(Base.julia_cmd()) --precompiled=yes`
+    # --startup-file
+    let JL_OPTIONS_STARTUPFILE_ON = 1,
+        JL_OPTIONS_STARTUPFILE_OFF = 2
+        # `HOME=/tmp` to avoid errors in the user .juliarc.jl, which hangs the tests.  Issue #
```
Comment (tkelman) [inline test/cmdlineargs.jl]: add the actual issue number?

Comment (rsrock) [inline test/cmdlineargs.jl]: That would be good to have, wouldn't it...

**Comment by vtjnash** [comment]
lgtm

**Comment by rsrock** [comment]
The issue number is in (thanks!), so I think this is ready

**Pull request merged**